0000-ldap-sync-tool.md

```diff
@@ -0,0 +1,233 @@
+
+The `mgr-ldapsync` should distinguish between LDAP users and Uyuni
+users, and operate adding, modifying or removing them only in LDAP
+context. As long as the user is not specified in LDAP, it should be
+treated as "Uyuni local users" and be ignored therefore.
+
+## Clashed UIDs
+
+Any UID that exists in the Uyuni server but has different
+authentication method than PAM or is mentioned as "frozen users"
+should be marked as conflicting and therefore ignored.
+
+Conflicting UIDs should be not processed, but reported to the log file
+as erroneous. However this should not prevent to process the rest
+of the existing users.
+
+However, any UID that has PAM authentication mode configured or is not
+mentioned as "frozen user" should be managed by LDAP.
+
+## Usage
+
+Typical use of the `mgr-ldapsync` should be straightforward by
+periodically run in cron. In console mode it should also provide a
+possibility of "dry-run" where only planning of actions to be taken
+will be shown.
+
+The output should be stored either to a configured log file or to the
+STDOUT, if requested.
+
+# Uyuni UI Enhancements
+
+Considering that this tool a new way of creating/modifying users
+beside the UI and API in an automated way, there should be an API
+endpoint that allows to set a flag in User Management that gives
+admins the clear indication that the certain user is managed externally.
+
+Main reason is the role mapping that is included. An admin currently
+easily can change the user settings in the UI and will be wondering
+why his changes get reset to the original settings. Therefore, Uyuni
+UI should mark external users either with special icon or color etc
+and the page of such user should have disabled "Update" button and an
+explanation for such user is managed elsewhere externally.
+
+# Reference Implementation
+
+This is already implemented in the prototype which is available at 
+[uyuni-ldap-sync repository](https://github.com/isbm/uyuni-ldap-sync).
+
+Current prototype has two releases:
+1. [v0.1](https://github.com/isbm/uyuni-ldap-sync/releases/tag/v0.1) just reimplements
+[existing script in Uyuni](https://github.com/uyuni-project/uyuni/blob/master/utils/sw-ldap-user-sync).
+2. [v0.2](https://github.com/isbm/uyuni-ldap-sync/releases/tag/v0.2) implements the current RFC and adds Active Directory support
+
+Moving from prototype status to a production release, the following
+would be still required:
+
+- Add unit tests
+- Possible refactorings
+
+## Additional Features
+
+Features that might come with the next release:
+
+- Automatic (transparent) PAM authentication setup/un-setup
+- Daemon mode
+
+# Drawbacks
+[drawbacks]: #drawbacks
+
+N/A
+
+# Alternatives
+[alternatives]: #alternatives
+
+## Native Implementation
+
+Alternative design would be to extend user management in the very
+Uyuni server, making it LDAP-aware natively. However, the amount of
+changes to the existing code might be very large.
+
+## Salt-based Implementation
```

Comment:
Does this alternative have any overlap with https://github.com/uyuni-project/uyuni-rfc/blob/master/accepted/00061-hub-users.md ?

Reply:
Frankly, I was considering about doing it with Salt at the beginning. I was also considering having this LDAP sync as an Ansible style module one can just call via Salt Ansible Gate (btw, still possible to add that interface).  The https://github.com/uyuni-project/uyuni-rfc/blob/master/accepted/00061-hub-users.md is that alternative. 

The difference between current RFC is that it is a daemon performing on its own, transparently reflecting changes in LDAP, without extra intervention and manually spawn Salt every time one wants to update LDAP data. Putting such Salt job to a cron would be an overkill and would introduce other issues (two minions issue, jobs racing etc). Also, Salt can just call this tool or we can have "Ansible module mode", so it will return output in Ansible-compatible JSON protocol and Salt can leverage that via earlier mentioned Ansible Gate subsystem.

I personally think that "Salt calls a tool Foo" is a much better aproach than "Salt implements Foo in Formula on its own" from reliability, testing and debug-ability perspectives.

Reply:
> Putting such Salt job to a cron looks like an overkill and might even introduce additional issues we could just avoid by design.

What issues can you anticipate?

> firing Salt [...with...] periodical crontab [...] is probably not the best idea [...]. This makes little to none sense to me [...]

I hear you when you mention that, from a software design perspective, a standalone daemon could look better than `cron` + `salt`.

Do you see additional reasons for this idea to be superior, on top of the simpler design?

> Salt [...] can just call this tool from a runner [...] I personally think that "Salt calls a tool Foo" is a much better aproach than "Salt implements Foo in Formula on its own"

OK, but that approach is not really proposed in this RFC, so while I could agree on the general idea, it is not really being discussed here.

> Last and not least, Salt [`pillar_ldap`](https://docs.saltstack.com/en/latest/ref/pillar/all/salt.pillar.pillar_ldap.html#configuration) does not support attribute and objects remapping. 

So, IIUC, with RFC 61 users are supposed to specify the mapping from LDAP attributes to Uyuni configuration (Users/Groups/permissions and so on) via Salt templates (Jinja, Python, etc.). With this RFC, the tool would have to include additional configuration options (worst case: a complete "mini-language") to specify that mapping - complexity would depend on how complicated the mapping needs to be. Would that be accurate?

Thanks

Reply:
> Do you see additional reasons for this idea to be superior, on top of the simpler design?

Thank you for the compliment that my design is simpler!

Yes, sure:

- Easier to debug than formulas. App  has full-blown logging facility built-in through the code. Salt Formulas, however, one would need to [modify Formula adding print-style logging](https://docs.saltstack.com/en/latest/topics/jinja/index.html#debugging) in order to narrow down the possible problem. Partially compiled state might only add to the  confusion and could increase debugging efforts.
- Possibility to clearly outline what is supported and what is not, keeping users from randomly modifying things. Feature request for missing parts.
- Keeps it _configurable_ at configuration files vs  _implementable_ in Jinja formulas. Salt is for configuration, daemon is for delivery. My approach forces you to keep things simple and segregated: let Salt deploy the configuration and run the daemon. It would make more sense from Salt perspective to do this:

```yaml
acme:
  ldapsync.running
    uyuni:
      user: pillar ......
    directory:
      cn: .......
```

> So, IIUC, with RFC 61 users are supposed to specify the mapping from LDAP attributes to Uyuni configuration (Users/Groups/permissions and so on) via Salt templates (Jinja, Python, etc.). With this RFC, the tool would have to include additional configuration options (worst case: a complete "mini-language") to specify that mapping - complexity would depend on how complicated the mapping needs to be. Would that be accurate?

Yes (reimplement sync in pure Jinja)  and no (add mini language for remap):

1. I could hardly call that "configuration". I would rather call it _full-blown implementation_. Implementation, one is going to debug. Debug at the Formula level. Worse than that, due to deviation from LDAP standard schema (or existing of POSIX and AD at first place), there consequently will be at least few if not more of those sets of Jinja templates with some slight subtle changes (read: more reported bugs). Bugs, that are only fixable to _that particular custom case_ and not applicable elsewhere. So this brings also whole new layer of versioning difficulty overhead. Therefore ideally one would like to rather implement LDAP sync itself in a separate Salt module, operating only data structures from Salt Formulas. This would allow keeping Jinja Formulas separated, taking care only of remappings/etc (again, this additional overhead only due to `pillar_ldap` limitations) and thus the Formula would format some commonly accepted structure that the module we are talking about would receive it then sync with the Uyuni. But such setup I find fragile and complicated.

2. No mini language is needed for remapping. Remapping is [already implemented as per RFC](https://github.com/isbm/uyuni-ldap-sync/commit/4123b5de2783ed9016544089bb35f47d534cda2b). You need to just write CN, proper LDAP filters and manually map custom objects to POSIX or AD in the config. This seems also _often_ needed, since admins tend to often deviate from the standard LDAP schemas writing their own stuff. Mini language will be actually required in Formulas: having additional couple of loops to figure out what to remap with before syncing.

3. My approach is basically alongside how [`slapd.conf`](https://linux.die.net/man/5/slapd.conf) does and the feature-set I am trying to reach how [Novell Filr](https://www.novell.com/documentation/novell-filr-1-2/filr-1-2_admin/data/ldap.html) does.

4. User `present`/`absent`/`updated` actions are still **insufficient** to properly sync users. Still somehow Salt Formula needs to know which user is local _prior to the sync_ and thus "untouchable". Same what Novell Filr does, BTW. So if there is a user with the ID `john` in LDAP with the role "Read Only" and the same _local user UID_ is in Uyuni server as `org_admin`, then you are asking for a trouble. You also cannot wipe out everyone from Uyuni, because you need to login to XML-RPC using local user that has `org_admin` perms, so at least one user needs to be reserved, which I foresee a hard-coded UID in the Formulas.

Hopefully this helps.